src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java

```diff
@@ -316,40 +332,136 @@ public void insertBufferedRows() {
     if (intermediateBuffer.isEmpty()) {
       LOGGER.info("No Rows Buffered, returning;");
       this.previousFlushTimeStampMs = System.currentTimeMillis();
-      return;
+      return null;
     }
-    LOGGER.debug(
-        "Invoking insertRows API for channel:{}, noOfRecords:{}, startOffset:{}, endOffset:{}",
-        this,
+
+    InsertValidationResponse response =
+        insertRowsWithFallback((StreamingBuffer) intermediateBuffer);
+    // Updates the flush time (last time we called insertRows API)
+    this.previousFlushTimeStampMs = System.currentTimeMillis();
+
+    LOGGER.info(
+        "[INSERT_ROWS] Successfully called insertRows for channel:{}, noOfRecords:{},"
+            + " startOffset:{}, endOffset:{}, hasErrors:{}",
+        this.getChannelName(),
         intermediateBuffer.getNumOfRecord(),
         intermediateBuffer.getFirstOffset(),
-        intermediateBuffer.getLastOffset());
-    InsertValidationResponse response =
-        channel.insertRows(
-            (Iterable<Map<String, Object>>) intermediateBuffer.getData(),
-            Long.toString(intermediateBuffer.getLastOffset()));
+        intermediateBuffer.getLastOffset(),
+        response.hasErrors());
+    if (response.hasErrors()) {
+      handleInsertRowsFailures(response.getInsertErrors(), intermediateBuffer.getSinkRecords());
+    }
+    return response;
+  }
 
-    // need to update the flush time(last I called insertrows)
-    this.previousFlushTimeStampMs = System.currentTimeMillis();
+  /**
+   * Uses {@link Fallback} API to reopen the channel if insertRows throws {@link SFException}.
+   *
+   * <p>We have deliberately not performed retries on insertRows because it might slow down overall
+   * ingestion and introduce lags in committing offsets to Kafka.
+   *
+   * <p>Note that insertRows API does perform channel validation which might throw SFException if
+   * channel is invalidated.
+   *
+   * <p>It can also send errors {@link
+   * net.snowflake.ingest.streaming.InsertValidationResponse.InsertError} in form of response inside
+   * {@link InsertValidationResponse}
+   *
+   * @param buffer buffer to insert into snowflake
+   * @return InsertValidationResponse object sent from insertRows API.
+   */
+  private InsertValidationResponse insertRowsWithFallback(StreamingBuffer buffer) {
+    Fallback<InsertValidationResponse> reopenChannelFallbackExecutorForInsertRows =
+        Fallback.builder(() -> insertRowsFallbackSupplier(buffer))
+            .handle(SFException.class)
```

Comment:
I do want to refactor common logic for retry into another class but I will do that after PrPr. 

Based on chat with @sfc-gh-mnaides and @sfc-gh-tjones only SFExceptions are worth retrying. I generally dont like the idea of catch all exceptions because that essentially will hide away other problems.

Reply:
We will have logs on which of two operations (open channel or insertRows) failed. This logger is kind of useful to know that even fallback has failed.